src/java.base/share/classes/java/util/HashMap.java

```diff
@@ -495,9 +495,9 @@ final void putMapEntries(Map<? extends K, ? extends V> m, boolean evict) {
         int s = m.size();
         if (s > 0) {
             if (table == null) { // pre-size
-                float ft = ((float)s / loadFactor) + 1.0F;
-                int t = ((ft < (float)MAXIMUM_CAPACITY) ?
-                         (int)ft : MAXIMUM_CAPACITY);
+                double dt = Math.ceil(s / loadFactor);
```

Comment:
Quick update on this. The division must be done in double, not float, before being passed to `Math.ceil`. I'd recommend casting the load factor to double:

    double dt = Math.ceil(s / (double) loadFactor);

The reason is that the significand of a float has only 24 bits, so the low order 8 bits of the eventual `int` will end up being zeroes even when they have some significant values. This starts to occur at larger sizes. For example, if `s` is 25165825, the result is 33554434.0 with double division but it's only 33554432.0 with float division. This latter value is too small, so if it were used the HashMap would need to be resized when `s` mappings are added.

Similar changes should be made in `readObject` below and in `WeakHashMap`.

I haven't had a chance to look at the test yet, sorry. You might want think about how to test for cases like this. Don't go off the deep end though. I don't think we want to have a test that creates lots of maps with 25 million entries, and we also don't want to duplicate the computations in the test itself like the Enum ConstantDirectory test does.